core/domain/suggestion_registry.py

```diff
@@ -319,30 +318,124 @@ def accept(self, commit_message):
+        """Validates a suggestion object of type SuggestionAddQuestion.
+
+        Raises:
+            ValidationError: One or more attributes of the SuggestionAddQuestion
+                object are invalid.
+        """
+        super(SuggestionAddQuestion, self).validate()
+
+        if self.get_score_type() != suggestion_models.SCORE_TYPE_QUESTION:
+            raise utils.ValidationError(
+                'Expected the first part of score_category to be %s '
+                ', received %s' % (
+                    suggestion_models.SCORE_TYPE_QUESTION,
+                    self.get_score_type()))
+        if not isinstance(self.change_cmd, question_domain.QuestionChange):
+            raise utils.ValidationError(
+                'Expected change_cmd to be an instance of QuestionChange')
+
+        if not self.change_cmd.cmd:
+            raise utils.ValidationError('Expected change_cmd to contain cmd')
+
+        if (
+                self.change_cmd.cmd !=
+                question_domain.CMD_ADD_QUESTION_FROM_SUGGESTION):
+            raise utils.ValidationError('Expected cmd to be %s, obtained %s' % (
+                question_domain.CMD_ADD_QUESTION_FROM_SUGGESTION,
+                self.change_cmd.cmd))
+
+        if not self.change_cmd.question_dict:
+            raise utils.ValidationError(
+                'Expected change_cmd to contain question_dict')
+
+        if 'question_state_data' not in self.change_cmd.question_dict:
+            raise utils.ValidationError(
+                'Expected question dict to contain question_state_data')
+
+        if 'language_code' not in self.change_cmd.question_dict:
+            raise utils.ValidationError(
+                'Expected question dict to contain language_code')
+
+        if (
+                'question_state_schema_version' not in
+                self.change_cmd.question_dict):
+            raise utils.ValidationError(
+                'Expected question dict to contain'
+                ' question_state_schema_version')
+
+        if not self.change_cmd.skill_id:
+            raise utils.ValidationError(
+                'Expected change_cmd to contain skill_id')
+
+    def pre_accept_validate(self):
+        """Performs referential validation. This function needs to be called
+        before accepting the suggestion.
         """
-        suggestion = cls(
-            suggestion_dict['suggestion_id'],
-            suggestion_dict['target_id'],
-            suggestion_dict['target_version_at_submission'],
-            suggestion_dict['status'], suggestion_dict['author_id'],
-            suggestion_dict['final_reviewer_id'], suggestion_dict['change_cmd'],
-            suggestion_dict['score_category'], suggestion_dict['last_updated'])
+        question_dict = self.change_cmd.question_dict
+        self.validate()
+        if (
+                question_dict['question_state_schema_version'] !=
+                feconf.CURRENT_EXPLORATION_STATES_SCHEMA_VERSION):
```

Comment:
/cc @aks681 -- is this the same for both states in questions and states in explorations? If so, should we call it STATE_SCHEMA_VERSION instead? If not, then shouldn't this be CURRENT_QUESTION_STATE_SCHEMA_VERSION?

Need to think through this and get it right -- getting it wrong has serious consequences. On looking at the code, it looks like Question depends on State. So consider the case where the State object is upgraded. This seems to imply the following things:

(a) It is correct to use feconf.CURRENT_EXPLORATION_STATES_SCHEMA_VERSION here.
(b) Questions should automigrate to the latest state schema version when loaded.
(c) The question data schema version should track feconf.CURRENT_EXPLORATION_STATES_SCHEMA_VERSION rather than a separate feconf.CURRENT_QUESTION_SCHEMA_VERSION.

Is that correct? If so, this is a no-op for @nithusha21, but @aks681 I think we need to file an issue to fix (b) and (c) ASAP. 

Also /cc @WickedBrat.

Reply:
Yeah, I had also thought about this. Basically, states in explorations and state data in questions refer to the same schema, so they have the same version constant. So, yeah, renaming to STATE_SCHEMA_VERSION sounds right.
Yeah, there should be a migration function to migrate question state data to the latest version. I think the same one declared in exp_services and exp_domain can be used for this purpose.
Also, a one-off job has to be created to handle question state data migrations. Modified this [issue](https://github.com/oppia/oppia/issues/5052) to include that.

On the same lines, should we, at some point, move the class State out of exp_domain, since it is now used by Question also?

Reply:
Yes, that sounds right -- and we should make sure that questions uses that constant instead of its own thing. That's pretty important.

Agree that one-off job needs to be created. But also we need to support on-the-fly migration to latest version (even before we run the migration job), just like for explorations. That's more important actually.

And yes, moving State out of exp_domain sounds good.

Could you please ensure these are tracked in issues (separating them where possible, but keeping together any stuff that needs to be done together?)